Docs/simplify index pages (#3995)
514-labs/moose: apps/framework-docs-v2/content/moosestack/olap/index.mdx

---
title: Moose OLAP

> description: Understand what Moose OLAP manages, when to use it, and where to start with ClickHouse

Reviewer: I think the previos active verb was better.
```suggestion
description: Create and manage ClickHouse tables with Moose OLAP, understand when to use Moose OLAP, and where to start with ClickHouse
```

# Moose OLAP

Moose OLAP is MooseStack's schema-as-code layer for ClickHouse. It turns tables, views, and materialized views into typed objects your app can read from, write to, and evolve in code.

<LanguageTabs>
  <LanguageTabContent value="typescript">
```ts filename="app/table.ts"
import { OlapTable, ClickHouseEngines, LifeCycle } from "@514labs/moose-lib";

interface Record {
  id: string;
  name: string;
  timestamp: Date;
}

export const table = new OlapTable<Record>("table", {
  engine: ClickHouseEngines.MergeTree,
  orderByFields: ["id"],
  version: "0.0",
  lifeCycle: LifeCycle.FULLY_MANAGED,
});
```
  </LanguageTabContent>
  <LanguageTabContent value="python">
```py filename="app/table.py"
from datetime import datetime

from moose_lib import ClickHouseEngines, LifeCycle, OlapConfig, OlapTable
from pydantic import BaseModel

class Record(BaseModel):
    id: str
    name: str
    timestamp: datetime

table = OlapTable[Record]("table", config=OlapConfig(
  engine=ClickHouseEngines.MergeTree,
  order_by_fields=["id"],
  version="0.0",
  life_cycle=LifeCycle.FULLY_MANAGED,
))
```
  </LanguageTabContent>
</LanguageTabs>

> Moose OLAP is the foundation for the rest of MooseStack: you can start with ClickHouse schema management, then layer on [Streaming](/moosestack/streaming), [APIs](/moosestack/apis), or [Workflows](/moosestack/workflows) as your system grows.

Reviewer: then layer on [Streaming](/moosestack/streaming), [APIs](/moosestack/apis), or [Workflows](/moosestack/workflows) as your system grows.

Is this relevant here / now?

> ## What can you do with Moose OLAP?

Reviewer: I don't think this heading is needed
```suggestion
```

Moose OLAP gives you a few core capabilities around your ClickHouse schema.

### Define your schema

- **[Define tables](/moosestack/olap/model-table)** in TypeScript or Python
- **[Shape and precompute data](/moosestack/olap/model-view)** with views and **[materialized views](/moosestack/olap/model-materialized-view)**

### Evolve or adopt an existing schema

> - **Carry local schema changes forward to staging and production** with **[schema migrations](/moosestack/olap/apply-migrations)** during development and **[planned migrations](/moosestack/olap/planned-migrations)** for remote environments

Reviewer: Why all the bold suddenly? Not fixing that, but slight wording suggestion

```suggestion
- **Carry local schema changes forward to staging and production environments** with **[schema migrations](/moosestack/olap/apply-migrations)** during development and **[planned migrations](/moosestack/olap/planned-migrations)** for remote environments
```

- **[Work with existing ClickHouse tables](/moosestack/olap/external-tables)** and **[pull those schemas into your project](/moosestack/olap/db-pull)** when Moose is not the only system managing them

### Read and write data

- **[Query data](/moosestack/olap/read-data)** from your application code or APIs through typed table and view objects
- **[Insert data](/moosestack/olap/insert-data)** for ingestion flows, backfills, or batch writes

## Get Started

> Pick the path that matches how your ClickHouse schema is managed today.

Reviewer: This only really maps to "use existing CH", no?
